cmd/up/exec.go

```diff
@@ -16,27 +16,268 @@ package up
+}
+
+func newHybridExecutor(ctx context.Context, up *upContext) (*hybridExecutor, error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	envsGetter, err := newEnvsGetter(up)
+	if err != nil {
+		return nil, err
+	}
+
+	envs, err := envsGetter.getEnvs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hybridExecutor{
+		workdir: filepath.Join(wd, up.Dev.Workdir),
+		envs:    envs,
+	}, nil
+}
+
+func newSyncExecutor(up *upContext) *syncExecutor {
+	return &syncExecutor{
+		iface:      up.Dev.Interface,
+		remotePort: up.Dev.RemotePort,
+	}
+}
+
+type configMapEnvsGetterInterface interface {
+	getEnvsFromConfigMap(ctx context.Context, name string, namespace string, client kubernetes.Interface) ([]string, error)
+}
+
+type secretsEnvsGetterInterface interface {
+	getEnvsFromSecrets(context.Context) ([]string, error)
+}
+
+type imageEnvsGetterInterface interface {
+	getEnvsFromImage(string) ([]string, error)
+}
+
+type imageGetterInterface interface {
+	GetImageMetadata(string) (registry.ImageMetadata, error)
+}
+
+type secretsGetterInterface interface {
+	GetUserSecrets(context.Context) ([]types.Secret, error)
+}
+
+type configMapGetter struct {
+}
+type secretsEnvsGetter struct {
+	secretsGetter secretsGetterInterface
+}
+type imageEnvsGetter struct {
+	imageGetter imageGetterInterface
+}
+
+type envsGetter struct {
+	dev                 *model.Dev
+	name, namespace     string
+	client              kubernetes.Interface
+	configMapEnvsGetter configMapEnvsGetterInterface
+	secretsEnvsGetter   secretsEnvsGetterInterface
+	imageEnvsGetter     imageEnvsGetterInterface
+}
+
+func newEnvsGetter(up *upContext) (*envsGetter, error) {
+
+	var secretsGetter secretsGetterInterface
+	if okteto.IsOkteto() {
+		oc, err := okteto.NewOktetoClient()
+		if err != nil {
+			return nil, err
+		}
+		secretsGetter = oc.User()
+	}
+
+	return &envsGetter{
+		dev:                 up.Dev,
+		name:                up.Manifest.Name,
+		namespace:           up.Manifest.Namespace,
+		client:              up.Client,
+		configMapEnvsGetter: &configMapGetter{},
+		secretsEnvsGetter: &secretsEnvsGetter{
+			secretsGetter: secretsGetter,
+		},
+		imageEnvsGetter: &imageEnvsGetter{
+			imageGetter: registry.NewOktetoRegistry(okteto.Config{}),
+		},
+	}, nil
+}
+
+func (eg *envsGetter) getEnvs(ctx context.Context) ([]string, error) {
+	var envs []string
+
+	configMapEnvs, err := eg.configMapEnvsGetter.getEnvsFromConfigMap(ctx, eg.name, eg.namespace, eg.client)
+	if err != nil {
+		return nil, err
+	}
+	envs = append(envs, configMapEnvs...)
+
+	secretsEnvs, err := eg.secretsEnvsGetter.getEnvsFromSecrets(ctx)
+	if err != nil {
+		return nil, err
+	}
+	envs = append(envs, secretsEnvs...)
+
+	app, err := apps.Get(ctx, eg.dev, eg.namespace, eg.client)
+	if err != nil {
+		return nil, err
+	}
+
+	imageEnvs, err := eg.imageEnvsGetter.getEnvsFromImage(app.PodSpec().Containers[0].Image)
```

Comment:
We shouldn't be checking `app.PodSpec().Containers[0].Image`. We can use the function `GetDevContainer` to get the container with the container name.

Reply:
done dd202d86cf1ce42babc90f207f706c127e0a4170